PYME/recipes/pointcloud.py

```diff
@@ -381,3 +384,98 @@ def execute(self, namespace):
             pass
         
         namespace[self.outputName] = res
+
+
+@register_module('BICGaussianMixtureModel')
+class BICGaussianMixtureModel(ModuleBase):
+    """Fit a Gaussian Mixture to a pointcloud, varying the number of components
+    and using the Bayesian Information Criterion to determine the optimum
+    number to use in the final model.
+
+    Parameters
+    ----------
+    input_points: PYME.IO.tabular
+        points to fit. Currently hardcoded to use x, y, and z keys.
+    n_max: Int
+        maximum number of Gaussians in the model
+    covariance: Enum
+        type of covariance to use in the model
+    output_labeled: PYME.IO.tabular
+        input source with additional column 'gmm_label' indicating predicted 
+        component membership of each point
+    """
+    input_points = Input('input')
+    n_max = Int(1)
+    covariance = Enum(('full', 'tied', 'diag', 'spherical'))
+    output_labeled = Output('labeled_points')
+    def execute(self, namespace):
+        from sklearn.mixture import GaussianMixture
+        from PYME.IO import MetaDataHandler
+
+        points = namespace[self.input_points]
+        X = np.stack([points['x'], points['y'], points['z']], axis=1)
+
+        n_components = range(1, self.n_max + 1)
+        bic = np.zeros(len(n_components))
+        for ind in range(len(n_components)):
+            gmm = GaussianMixture(n_components=n_components[ind], 
+                                  covariance_type=self.covariance)
+            gmm.fit(X)
+            bic[ind] = gmm.bic(X)
+            logger.debug('%d BIC: %f' % (n_components[ind], bic[ind]))
+
+        best = n_components[np.argmin(bic)]
```

Comment:
TODO - warn if the best fit is the highest tried n_components (or if excessively noisy)